Reduce cast usage for COUNT aggregate and add support for Mssql count_big

maca88 opened the pull request:
This is a continuation of #2036, which extends the cast reduce usage for count aggregate function. In addition, this PR introduces `count_big` sql function that is used when using `LongCount` and `MsSql2000Dialect` overrides it to use the specific Sql Server `count_big` function (by default `count` is used).

hazzik commented on src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs:
```diff
@@ -584,5 +601,118 @@ protected HqlTreeNode VisitNewArrayExpression(NewArrayExpression expression)
+				: persister.EntityMetamodel.PropertyTypes[index.Value];
+		}
+
+		private string TryGetEntityName(MemberExpression memberExpression)
```
This is not entirely correct.

maca88 replied:
You are right, I forgot about custom entity names. This method was "enhanced" and moved to `ExpressionsHelper` class as it is also used by #1996, #2036, #2079, but was not updated here. Nonetheless, both methods do not work when having custom entity names. I realized that `IEntityNameProvider` exists and is implemented by the linq query which I will use to calculate the correct entity name instead of using `TryGetGuessEntityName` method. I will update the mentioned PRs when the method is fixed.

maca88 replied:
I took me more than expected as there are a lot mapping features that the method needs to be aware of in order to travese them, but as far as I know the method is now as accurate as it could be. Unfortunately it still needs to guess the entity name when casting is involved (like `((PizzaOrder)o.Order).Name`), but there is nothing we can do about it. I 've added a test for entity names in #2054. 
As this PR has an extension for `ISQLFunction` similar to what am I planning to do in #2061, I will update it after the other one is finished.

maca88 commented:
Rebased and renamed to `ISQLFunction` extension interface as it will be used in all `ISQLFunction` implementations which I will address in a separate PR. Initially, I planned to deprecate `ISQLFunction.ReturnType` in favor of the newly added `ISQLFunctionExtended.GetEffectiveReturnType` method but it turned out that it would produce an unwanted breaking change. For example the following test would fail in Sql Server:
https://github.com/nhibernate/nhibernate-core/blob/d4206e24312b323e1da2f49affc0225d05866a3c/src/NHibernate.Test/QueryTest/CountFixture.cs#L23-L24
because `count` effective return type in Sql Server is an `int`. So the idea is that `ISQLFunction.ReturnType` should return the same type no matter what database is used, which is basically what it does today, except for some functions which IMO should be corrected:
- length: in Firebird and Oracle dialects returns `Int64`, SysbaseASE9 returns `string` (very suspicious) and in other dialect returns `Int32`.
- mod: in Firebird dialect returns `Double` where in other dialects returns `Int32`

The intention of `ISQLFunctionExtended.GetEffectiveReturnType` method is be used by higher level api like linq in order to determine whether a cast or a transparent cast is needed.

maca88 commented on src/NHibernate/Dialect/MsSql2000Dialect.cs:
```diff
@@ -286,7 +286,8 @@ protected virtual void RegisterKeywords()
 		{
-			RegisterFunction("count", new CountBigQueryFunction());
+			RegisterFunction("count", new CountQueryFunction());
+			RegisterFunction("count_big", new CountBigQueryFunction());
```
> Isn't this still causing the breaking change you were writing about?

For HQL nothing is changed, the new `GetEffectiveReturnType` method is used only for Linq queries. `CountQueryFunction.ReturnType` still returns `Int64`, which is what HQL uses to determine the type, so hql queries using `count` will always be casted to `Int64`.

maca88 commented on src/NHibernate/Hql/Ast/ANTLR/SqlGenerator.g:
```diff
@@ -150,7 +150,7 @@ selectExpr
 
 count
-	: ^(COUNT { Out("count("); }  ( distinctOrAll ) ? countExpr { Out(")"); } )
+	: ^(c=COUNT { OutAggregateFunctionName(c); Out("("); }  ( distinctOrAll ) ? countExpr { Out(")"); } )
```
>  I mean, was this causing the SQL-Server HQL count to count_big mapping to be ignored by this piece of code, causing NHibernate to anyway use SQL-Server int32 count?

Correct, `count` was hardcoded, which was the reason why changing from `count` to `count_big` in `CountBigQueryFunction` didn't work.

maca88 commented on src/NHibernate/Dialect/Function/ISQLFunction.cs:
```diff
@@ -41,4 +43,47 @@ public interface ISQLFunction
+				{
+#pragma warning disable 618
+					return sqlFunction.ReturnType(argumentTypes.FirstOrDefault(), mapping);
+#pragma warning restore 618
```
Correct, they will be needed in #2316 instead. Rebased and removed the pragma warning.

hazzik commented on src/NHibernate/Dialect/Function/ISQLFunctionExtended.cs:
```diff
@@ -0,0 +1,27 @@
+		/// <summary>
+		/// The function name or <see langword="null"/> when multiple functions/operators/statements are used.
+		/// </summary>
+		string FunctionName { get; }
```
I think this should be renamed to `Name` to reduce need of obsoleting things.

maca88 replied:
Why not, renamed in #2359.

hazzik commented on src/NHibernate/Hql/Ast/ANTLR/Tree/AggregateNode.cs:
```diff
@@ -19,6 +20,19 @@ public AggregateNode(IToken token)
 		{
 		}
 
+		public string FunctionName
```
I think this could be an overgeneralization and should be refactored into a node-specific code unless you see that it can be extended for functions other than `count` and `count_big`

maca88 replied:
This may be extended for string concatenation, which is supported by some databases ([Sql server](https://docs.microsoft.com/en-us/sql/t-sql/functions/string-agg-transact-sql?view=sql-server-ver15), [MySql](https://dev.mysql.com/doc/refman/8.0/en/group-by-functions.html#function_group-concat), [Postgresql](https://www.postgresql.org/docs/9.1/functions-aggregate.html) and [Oracle](https://docs.oracle.com/database/121/SQLRF/functions101.htm#SQLRF30030)), where different function names are used (`LISTAGG`, `GROUP_CONCAT`, `STRING_AGG`).

Nazgaul replied:
@maca88 can you provide some sort of prototype of string_agg? I'm dying to use this with linq provider

maca88 replied:
To my knowledge it is not possible to implement it without modifying NHibernate source code. It is on my todo list for the next release, so unfortunately you will have to wait.
